### PR TITLE
Better overflow detection for AssignAdd/AssignSub

### PR DESCRIPTION
**Pull request by kevinbackhouse**
Previously, to detect overflow in `AssignAdd` or `AssignSub`, you had to use `defMightOverflow`, which was a bit inconvenient. This PR adds them to the set of analyzable expressions, so that you can just use `convertedExprMightOverflow`, like normal.

### CODE REVIEW

**Comment by geoffw0** [comment]
I've created some tests for this PR, hopefully we can get it merged soon.

File: cpp/ql/test/library-tests/rangeanalysis/SimpleRangeAnalysis/test.c [hunk 1]
```diff
@@ -398,3 +398,29 @@ unsigned int test_comma01(unsigned int x) {
...
+  out(i); // 30
+
+  i = j = 40;
+  out(i); // 40
+
+  i = (j += 10);
+  out(i); // 50
+
+  i = 20 + (j -= 10);
+  out(i); // 60
```
Comment (jbj) [inline cpp/ql/test/library-tests/rangeanalysis/SimpleRangeAnalysis/test.c]: Should that be `// 60 [BUG: the analysis thinks it's 2^-31 .. 2^31-1]`?

**Comment by geoffw0** [comment]
An existing test is failing because of two test lines that were previously detected by `ArithmeticOverflowTainted.ql`:
```
| test.c:44:7:44:10 | len2 | $@ flows to here and is used in arithmetic, potentially causing an underflow. | test.c:41:17:41:20 | argv | User-provided value |
| test.c:54:7:54:10 | len3 | $@ flows to here and is used in arithmetic, potentially causing an underflow. | test.c:51:17:51:20 | argv | User-provided value |
```
Now they are also detected by `IntegerOverflowTainted.ql`:
```
| test.c:44:7:44:12 | ... -- | $@ flows to here and is used in an expression which might overflow negatively. | test.c:41:17:41:20 | argv | User-provided value |
| test.c:54:7:54:12 | ... -- | $@ flows to here and is used in an expression which might overflow negatively. | test.c:51:17:51:20 | argv | User-provided value |
```
Are we OK with simply accepting the duplication here?

**Comment by jbj** [comment]
I assume you mean `ArithmeticTainted.ql`. I'm okay with this duplication as long as both of these queries are hidden from LGTM -- they're "warning-low". Users that use the whole security suite will have different expectations.

**Comment by geoffw0** [comment]
Then please merge my PR https://github.com/kevinbackhouse/ql/pull/2 and after the tests pass I will merge this.  I should also perform 'superficial performance testing'...

**Comment by geoffw0** [comment]
I performance tested this with `ComparisonWithCancellingSubExpr.ql`, `SnprintfOverflow.ql` and `IntegerOverflowTainted.ql`.  Performance is a little slower (around 5% consistently).  It's difficult to analyse what's going on thanks to some complicated recursion, but I'm guessing this is simply the result of there being more analysable expressions:
```
	IntegerOverflowTainted.ql-14:SimpleRangeAnalysis::getUpperBoundsImpl#ff ........................................... 56.7s (executed 3066 times)
	IntegerOverflowTainted.ql-14:SimpleRangeAnalysis::getLowerBoundsImpl#ff ........................................... 36.9s (executed 1945 times)
	IntegerOverflowTainted.ql-14:SimpleRangeAnalysis::getDefUpperBoundsImpl#fff ....................................... 29.9s (executed 2800 times)
	IntegerOverflowTainted.ql-14:SimpleRangeAnalysis::getDefLowerBoundsImpl#fff ....................................... 23s (executed 2192 times)
	IntegerOverflowTainted.ql-17:Expr::Expr::toString_dispred#ff ...................................................... 16.9s
	IntegerOverflowTainted.ql-13:idominance@Dominance::functionEntry#1@BasicBlocks::Cached::bb_successor_cached#2#fff . 16.8s
	IntegerOverflowTainted.ql-14:SimpleRangeAnalysis::getDefUpperBounds#fff ........................................... 16.5s (executed 1640 times)
	IntegerOverflowTainted.ql-9:ConstantExprs::reachableRecursive#f ................................................... 10.6s (executed 3703 times)
	IntegerOverflowTainted.ql-14:SimpleRangeAnalysis::getDefLowerBounds#fff ........................................... 10.3s
```
(qt, total 772s)
becomes:
```
	IntegerOverflowTainted.ql-14:SimpleRangeAnalysis::getUpperBoundsImpl#ff ........................................... 1m21s (executed 3069 times)
	IntegerOverflowTainted.ql-14:SimpleRangeAnalysis::getLowerBoundsImpl#ff ........................................... 51.9s (executed 1946 times)
	IntegerOverflowTainted.ql-14:SimpleRangeAnalysis::getDefUpperBoundsImpl#fff ....................................... 31.5s (executed 2801 times)
	IntegerOverflowTainted.ql-14:SimpleRangeAnalysis::getDefLowerBoundsImpl#fff ....................................... 23.5s (executed 2192 times)
	IntegerOverflowTainted.ql-14:SimpleRangeAnalysis::getDefUpperBounds#fff ........................................... 16.9s (executed 1640 times)
	IntegerOverflowTainted.ql-13:idominance@Dominance::functionEntry#1@BasicBlocks::Cached::bb_successor_cached#2#fff . 16.5s
	IntegerOverflowTainted.ql-17:Expr::Expr::toString_dispred#ff ...................................................... 15.2s
	IntegerOverflowTainted.ql-9:ConstantExprs::reachableRecursive#f ................................................... 10.5s (executed 3703 times)
	IntegerOverflowTainted.ql-14:SimpleRangeAnalysis::getDefLowerBounds#fff ........................................... 10.4s
```
(qt, total 816s)

**Pull request merged**